# Nemovitosti – ETL Pipeline

Alternativa k SQL skriptům `02_import_raw.sql` + `03_transform.sql`.  
Čte data přímo z xlsx, parsuje GPS a importuje do MySQL přes SQLAlchemy.

**Předpoklad:** `01_schema.sql` musí být spuštěn (databáze a tabulky existují).

### Závislosti
```
pip install pandas openpyxl sqlalchemy pymysql cryptography requests
```
*Instalace knihoven, nastavení proměnných a databázové connection*

In [1]:
from pathlib import Path
import pandas as pd
import numpy as np
from sqlalchemy import create_engine, text, MetaData
from sqlalchemy.dialects.mysql import insert as mysql_insert

# ── Konfigurace ───────────────────────────────────────────────────────────────
DB_URL   = 'mysql+pymysql://boda:bodaboda@localhost/nemovitosti?charset=utf8mb4'
DATA_DIR = Path('../data')
CUZK_DIR = DATA_DIR / 'cuzk'

engine = create_engine(DB_URL, echo=False)
print('Připojeno:', engine.url.database)

Připojeno: nemovitosti


## 1. ČÚZK číselníky

---
*Načtení dat z CSV souboru s katastrálními úřady a pak katastrálními územími*

```
Kódování cp1250, oddělovač `;`. Načítáme jen sloupce relevantní pro projekt.

Vazba mezi těmito soubory je dle popisu na www stránkách df_urady.kod = df_uzemi.prares_kod
```

---

In [2]:
df_urady = pd.read_csv(
    CUZK_DIR / 'SC_PRACRES_DOTAZ.csv',
    encoding='cp1250', sep=';', dtype=str, keep_default_na=False
)
df_urady.columns = df_urady.columns.str.strip()

# Odstranění prázdných řetězců a nahrazení NaN
for col in df_urady.columns:
    df_urady[col] = df_urady[col].str.strip().replace('', np.nan)

df_urady = df_urady.rename(columns={
    'KOD': 'kod', 'ZKRATKA': 'zkratka', 'TYP_PRAC': 'typ_prac',
    'NAZEV': 'nazev', 'NAZEV_ZKRACENY': 'nazev_zkraceny',
    'NADRIZ_PRAC': 'nadriz_prac', 'PLATNOST_OD': 'platnost_od',
    'PLATNOST_DO': 'platnost_do', 'ICO': 'ico',
    'TELEFON': 'telefon', 'FAX': 'fax', 'EMAIL': 'email',
    'OBEC': 'obec', 'MESTSKY_OBVOD': 'mestsky_obvod',
    'CAST_OBCE': 'cast_obce', 'CISLO_DOMOVNI': 'cislo_domovni',
    'NAZEV_ULICE': 'nazev_ulice', 'CISLO_ORIENTACNI': 'cislo_orient',
    'PSC': 'psc',
})[
    ['kod', 'zkratka', 'typ_prac', 'nazev', 'nazev_zkraceny',
     'nadriz_prac', 'platnost_od', 'platnost_do', 'ico',
     'telefon', 'fax', 'email', 'obec', 'mestsky_obvod',
     'cast_obce', 'cislo_domovni', 'nazev_ulice', 'cislo_orient', 'psc']
]

# Převod datových typů - ze string na numerické a datum
for col in ['kod', 'typ_prac', 'nadriz_prac']:
    df_urady[col] = pd.to_numeric(df_urady[col], errors='coerce')
df_urady['platnost_od'] = pd.to_datetime(df_urady['platnost_od'], format='%d.%m.%Y', errors='coerce').dt.date
df_urady['platnost_do'] = pd.to_datetime(df_urady['platnost_do'], format='%d.%m.%Y', errors='coerce').dt.date

print(f'katastralni_urady: {len(df_urady)} řádků')
df_urady.head(2)

katastralni_urady: 117 řádků


,kod,zkratka,typ_prac,nazev,nazev_zkraceny,nadriz_prac,platnost_od,platnost_do,ico,telefon,fax,email,obec,mestsky_obvod,cast_obce,cislo_domovni,nazev_ulice,cislo_orient,psc
0,10,ČÚZK,7,Český úřad zeměměřický a katastrální,ČÚZK,NaN,1993-01-01,NaT,25712,284041111,NaN,cuzk@cuzk.gov.cz,Praha,Praha 8,Kobylisy,1800,Pod sídlištěm,9,18200
1,20,ČÚZK-SCD,3,Český úřad zeměměřický a katastrální - SCD,ČÚZK-SCD,10.0,1993-01-01,NaT,25712,284041111,NaN,cuzk@cuzk.gov.cz,Praha,Praha 8,Kobylisy,1800,Pod sídlištěm,9,18200


In [3]:
df_uzemi = pd.read_csv(
    CUZK_DIR / 'SC_SEZNAMKUKRA_DOTAZ.csv',
    encoding='cp1250', sep=';', dtype=str, keep_default_na=False
)
df_uzemi.columns = df_uzemi.columns.str.strip()

# Odstranění prázdných řetězců a nahrazení NaN
for col in df_uzemi.columns:
    df_uzemi[col] = df_uzemi[col].str.strip().replace('', np.nan)

df_uzemi = df_uzemi.rename(columns={
    'KU_KOD': 'ku_kod', 'KU_NAZEV': 'ku_nazev',
    'KRAJ_KOD': 'kraj_kod', 'KRAJ_NAZEV': 'kraj_nazev',
    'OKRES_KOD': 'okres_kod', 'OKRES_NAZEV': 'okres_nazev',
    'OBEC_KOD': 'obec_kod', 'OBEC_NAZEV': 'obec_nazev',
    'PLATNOST_OD': 'platnost_od', 'PLATNOST_DO': 'platnost_do',
    'PRARES_KOD': 'prares_kod', 'PRARES_NAZEV': 'prares_nazev',
})[
    ['ku_kod', 'ku_nazev', 'kraj_kod', 'kraj_nazev',
     'okres_kod', 'okres_nazev', 'obec_kod', 'obec_nazev',
     'platnost_od', 'platnost_do', 'prares_kod', 'prares_nazev']
]

# Převod datových typů - ze string na numerické a datum
for col in ['ku_kod', 'kraj_kod', 'okres_kod', 'obec_kod', 'prares_kod']:
    df_uzemi[col] = pd.to_numeric(df_uzemi[col], errors='coerce')
df_uzemi['platnost_od'] = pd.to_datetime(df_uzemi['platnost_od'], format='%d.%m.%Y', errors='coerce').dt.date
df_uzemi['platnost_do'] = pd.to_datetime(df_uzemi['platnost_do'], format='%d.%m.%Y', errors='coerce').dt.date

print(f'katastralni_uzemi: {len(df_uzemi)} řádků')
df_uzemi.head(2)

katastralni_uzemi: 13074 řádků


,ku_kod,ku_nazev,kraj_kod,kraj_nazev,okres_kod,okres_nazev,obec_kod,obec_nazev,platnost_od,platnost_do,prares_kod,prares_nazev
0,601527,Běchovice,19,Hlavní město Praha,NaN,NaN,554782,Praha,1994-03-01,NaT,101,"Katastrální úřad pro hlavní město Prahu, Katas..."
1,602582,Benice,19,Hlavní město Praha,NaN,NaN,554782,Praha,1991-10-01,NaT,101,"Katastrální úřad pro hlavní město Prahu, Katas..."


## 2. Nemovitosti
Čte přímo z xlsx (sheet `NEMOVITOSTI`). Oba datasety se sjednotí do jedné DataFrame.

In [4]:
COLS_OV = {
    'ID': 'id', 'Využití budovy': 'vyuziti_budovy',
    'Kraj': 'kraj', 'Okres': 'okres', 'MČ': 'mc',
    'Katastrální území': 'katastralni_uzemi',
    'Ulice č.p./č.e.': 'ulice_cp', 'Město': 'mesto', 'PSČ': 'psc',
    'GPS': 'gps_text', 'Vymezené byty': 'vymezene_byty',
    'Vlastnictví': 'vlastnictvi', 'Typ budovy': 'typ_budovy',
    'Ochrana': 'ochrana', 'Datum dokončení RUIAN': 'datum_dokonceni_ruian',
}
COLS_CD = {
    'ID': 'id', 'ID Subjektu': 'id_subjektu',
    'Využití budovy': 'vyuziti_budovy',
    'Kraj': 'kraj', 'Okres': 'okres', 'MČ': 'mc',
    'Katastrální území': 'katastralni_uzemi',
    'Ulice č.p./č.e.': 'ulice_cp', 'Město': 'mesto', 'PSČ': 'psc',
    'GPS': 'gps_text', 'Vymezené byty': 'vymezene_byty',
    'Vlastnictví': 'vlastnictvi', 'Typ budovy': 'typ_budovy',
    'Ochrana': 'ochrana',
}

# Funkce pro načítání a zpracování dat z Excelu. 
# Načte pouze sloupce, které jsou v mapě, a přidá sloupec s typem datasetu.
# Načte je jako string. Chybějící sloupce se ignorují. Načtená data se vrací jako DataFrame.
def load_xlsx(path, col_map, typ_datasetu):
    df = pd.read_excel(path, sheet_name='NEMOVITOSTI', dtype=str)
    df.columns = df.columns.str.strip()
    df = df.rename(columns=col_map)[[c for c in col_map.values() if c in df.rename(columns=col_map).columns]]
    df['typ_datasetu'] = typ_datasetu
    return df

df_ov = load_xlsx(DATA_DIR / 'Občanská vybavenost- nemovitosti.xlsx', COLS_OV, 'obcanska_vybavenost')
df_cd = load_xlsx(DATA_DIR / 'Činžovní domy - nemovitosti.xlsx',      COLS_CD, 'cinzovni_domy')

df = pd.concat([df_ov, df_cd], ignore_index=True)
print(f'Načteno: {len(df_ov)} + {len(df_cd)} = {len(df)} řádků')
df.head(2)

Načteno: 4760 + 2051 = 6811 řádků


,id,kraj,okres,mc,katastralni_uzemi,ulice_cp,mesto,psc,gps_text,vymezene_byty,vlastnictvi,typ_budovy,ochrana,datum_dokonceni_ruian,typ_datasetu,id_subjektu,vyuziti_budovy
0,879590101,Hlavní město Praha,Praha 2,Nové Město,Nové Město,Vodičkova 709,Praha 2 - Nové Město,11000,N 50.0815533703108 E 14.4243770057724,Ne,Ostatní,budova s číslem popisným,NaN,2000-12-31 00:00:00,obcanska_vybavenost,NaN,NaN
1,879607101,Hlavní město Praha,Praha 2,Nové Město,Nové Město,Václavské náměstí 841,Praha 2 - Nové Město,11000,N 50.0841224909442 E 14.4249750401906,Ne,Ostatní,budova s číslem popisným,NaN,2012-09-14 00:00:00,obcanska_vybavenost,NaN,NaN


## 3. Transformace

In [5]:
# Číselné ID
df['id'] = pd.to_numeric(df['id'], errors='coerce')
if 'id_subjektu' in df.columns:
    df['id_subjektu'] = pd.to_numeric(df['id_subjektu'], errors='coerce')

# Prázdné řetězce → None (dokud jsou sloupce ještě str dtype)
for col in df.select_dtypes(include='object').columns:
    df[col] = df[col].str.strip().replace({'': np.nan, 'NaT': np.nan, 'NaN': np.nan})

# Datum dokončení RUIAN – konverze až po string cleanup
if 'datum_dokonceni_ruian' in df.columns:
    df['datum_dokonceni_ruian'] = pd.to_datetime(
        df['datum_dokonceni_ruian'], format='%d.%m.%Y', errors='coerce'
    ).dt.date

# Inicializace geocoding sloupců
df['geocoding_shoda']  = np.nan
df['geocoding_adresa'] = np.nan

C:\Users\panak\AppData\Local\Temp\ipykernel_6256\1904291992.py:7: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for col in df.select_dtypes(include='object').columns:


In [6]:
# GPS parsing – formát: 'N 50.0815 E 14.4243'
def parse_gps(s):
    if pd.isna(s):
        return np.nan, np.nan
    s = str(s).strip()
    if not s.startswith('N ') or ' E ' not in s:
        return np.nan, np.nan
    try:
        parts = s.split(' E ')
        return float(parts[0][2:]), float(parts[1])
    except ValueError:
        return np.nan, np.nan

df[['gps_lat', 'gps_lon']] = df['gps_text'].apply(lambda x: pd.Series(parse_gps(x)))
print(f'GPS naparsováno: {df["gps_lat"].notna().sum()} / {len(df)} řádků')

GPS naparsováno: 6747 / 6811 řádků


In [7]:
# katastralni_uzemi_klic – strip textu za poslední pomlčkou
# Příklad: 'Říčany u Prahy - Říčany' → 'Říčany u Prahy'
def ku_strip(s):
    if pd.isna(s):
        return s
    s = str(s)
    return s[:s.rfind('-')].rstrip() if '-' in s else s

# 1. iterace
df['katastralni_uzemi_klic'] = df['katastralni_uzemi'].apply(ku_strip)

# 2. iterace – pro řádky stále nenapárované (např. 'Příbram - Příbram V-Zdaboř' → 'Příbram')
ku_names = set(df_uzemi['ku_nazev'].dropna()) # hodnoty z číselníku katastrálních území, pro rychlé hledání

# podmínka je - pokud katastrální území není v číselníku a zároveň obsahuje pomlčku, 
# zkusit znovu oddělit text za poslední pomlčkou
mask = (~df['katastralni_uzemi_klic'].isin(ku_names)) & \
        df['katastralni_uzemi_klic'].str.contains('-', na=False)
df.loc[mask, 'katastralni_uzemi_klic'] = df.loc[mask, 'katastralni_uzemi_klic'].apply(ku_strip)

matched = df['katastralni_uzemi_klic'].isin(ku_names).sum()
print(f'Napárováno na číselník: {matched} / {df["katastralni_uzemi_klic"].notna().sum()} řádků')

Napárováno na číselník: 6811 / 6811 řádků


In [8]:
# pocet_ov_7km – počet budov OV do 7 km od každého činžovního domu
# Pravoúhlá aproximace: d = sqrt((Δlat·111.32)² + (Δlon·111.32·cos(lat))²)  [km]

LAT_KM = 111.32

ov = df[
    (df['typ_datasetu'] == 'obcanska_vybavenost') &
    df['gps_lat'].notna() & df['gps_lon'].notna()
]
ov_lat = ov['gps_lat'].values
ov_lon = ov['gps_lon'].values

mask_cd = (
    (df['typ_datasetu'] == 'cinzovni_domy') &
    df['gps_lat'].notna() & df['gps_lon'].notna()
)

df['pocet_ov_7km'] = np.nan

pocet = []
for _, row in df[mask_cd].iterrows():
    cos_lat = np.cos(np.radians(row['gps_lat']))
    dlat_km = (ov_lat - row['gps_lat']) * LAT_KM
    dlon_km = (ov_lon - row['gps_lon']) * LAT_KM * cos_lat
    dist_km = np.sqrt(dlat_km**2 + dlon_km**2)
    pocet.append(int((dist_km <= 7.0).sum()))

df.loc[mask_cd, 'pocet_ov_7km'] = pocet
print(f'pocet_ov_7km spočítáno: {mask_cd.sum()} CD záznamů')
df.loc[mask_cd, 'pocet_ov_7km'].value_counts().sort_index().head(10)

pocet_ov_7km spočítáno: 2051 CD záznamů


pocet_ov_7km
0.0      2019
2.0         2
3.0        20
11.0        1
12.0        1
65.0        1
84.0        1
97.0        1
101.0       2
114.0       3
Name: count, dtype: int64

## 4. Validace

In [9]:
df['flag_bez_gps']   = df['gps_lat'].isna().astype(int)
df['flag_bez_ulice'] = df['ulice_cp'].isna().astype(int)
df['flag_bez_ku']    = (~df['katastralni_uzemi_klic'].isin(ku_names)).astype(int)

df.groupby('typ_datasetu')[['flag_bez_gps', 'flag_bez_ulice', 'flag_bez_ku']].sum()

,flag_bez_gps,flag_bez_ulice,flag_bez_ku
typ_datasetu,,,
cinzovni_domy,0,22,0
obcanska_vybavenost,64,1321,0


## 5. Import do MySQL

ČÚZK tabulky: TRUNCATE + INSERT.  
Nemovitosti: upsert přes `INSERT … ON DUPLICATE KEY UPDATE`  
(geocoding výsledky se při opakovaném importu nepřepíší).

In [10]:
# Import katastralni_urady a katastralni_uzemi do MySQL
meta = MetaData()
meta.reflect(bind=engine)

def to_none(v): # MySQL neakceptuje NaN ani NaT, musí být None
    """Převede float nan a pd.NaT na None pro MySQL."""
    try:
        return None if pd.isna(v) else v
    except (TypeError, ValueError):
        return v

def records(df): 
    return [{k: to_none(v) for k, v in row.items()} for row in df.to_dict('records')]

def truncate_and_load(engine, table_name, df):
    with engine.begin() as conn: # příprava - smazání dat a vypnutím FK kontrol
        conn.execute(text('SET foreign_key_checks = 0'))
        conn.execute(text(f'TRUNCATE TABLE {table_name}'))
        conn.execute(text('SET foreign_key_checks = 1'))
    with engine.begin() as conn: # vložení dat (do čisté tabulky, viz. výše)
        conn.execute(meta.tables[table_name].insert(), records(df))
    print(f'  {table_name}: {len(df)} řádků')

truncate_and_load(engine, 'katastralni_urady', df_urady)
truncate_and_load(engine, 'katastralni_uzemi', df_uzemi)

  katastralni_urady: 117 řádků
  katastralni_uzemi: 13074 řádků


In [11]:
DB_COLS = [
    'id', 'typ_datasetu', 'id_subjektu', 'vyuziti_budovy',
    'kraj', 'okres', 'mc', 'katastralni_uzemi', 'katastralni_uzemi_klic',
    'ulice_cp', 'mesto', 'psc', 'gps_text', 'gps_lat', 'gps_lon',
    'vymezene_byty', 'vlastnictvi', 'typ_budovy', 'ochrana',
    'datum_dokonceni_ruian', 'pocet_ov_7km',
]

NO_UPDATE = {'id', 'typ_datasetu', 'geocoding_shoda', 'geocoding_adresa', 'geocoding_metoda', 'created_at'}

def upsert_nemovitosti(engine, df, meta, chunk_size=500):
    table  = meta.tables['nemovitosti']
    export = df[[c for c in DB_COLS if c in df.columns]].copy()
    for col in DB_COLS:
        if col not in export.columns:
            export[col] = None
    data = records(export)

    with engine.begin() as conn:
        for i in range(0, len(data), chunk_size):
            chunk = data[i : i + chunk_size]
            stmt = mysql_insert(table).values(chunk)
            update = {
                col: stmt.inserted[col]
                for col in chunk[0]
                if col not in NO_UPDATE
            }
            conn.execute(stmt.on_duplicate_key_update(**update))

    print(f'  nemovitosti: {len(data)} řádků (upsert)')

upsert_nemovitosti(engine, df, meta)

  nemovitosti: 6811 řádků (upsert)


In [12]:
# Kontrola v MySQL – počet záznamů a základní statistika chybějících GPS, ulice, katastrálních území
with engine.connect() as conn:
    result = conn.execute(text('''
        SELECT typ_datasetu, COUNT(*) AS celkem,
               SUM(flag_bez_gps)   AS bez_gps,
               SUM(flag_bez_ulice) AS bez_ulice,
               SUM(flag_bez_ku)    AS bez_ku
        FROM  v_report
        GROUP BY typ_datasetu
    '''))
pd.DataFrame(result.fetchall(), columns=result.keys())

,typ_datasetu,celkem,bez_gps,bez_ulice,bez_ku
0,cinzovni_domy,2051,0,22,0
1,obcanska_vybavenost,4759,64,1321,4759


## 6. Geocoding – Mapy.cz API

Mapy.cz API má lepší kvalitu pro česká data než Nominatim.  
API klíč zdarma na [api.mapy.cz](https://api.mapy.cz).

| Buňka | Co dělá |
|-------|---------|
| Setup | API klíč + pomocné funkce |
| Forward geocoding | adresa → GPS pro záznamy s `flag_bez_gps = 1` |
| Reverse geocoding | GPS → adresa pro validaci, uloží do `geocoding_adresa` |

Výsledky se ukládají průběžně do DB — buňky lze spouštět nezávisle a opakovaně.

import subprocess, sys
subprocess.run([sys.executable, '-m', 'pip', 'install', 'requests'], check=True)

In [13]:
import requests, time

MAPY_API  = 'https://api.mapy.cz/v1'
API_KEY   = 'nYJ2gE48fnO2TASz1wwFRCGYVrTfvi331EpqhdwWRvA'   # viz https://api.mapy.cz → Dashboard → vygenerovat klíč
HEADERS   = {'Accept': 'application/json'}

def _parse_item(item):
    """Ze záznamu API vytáhne ulice (name), mesto (location), psc (zip), type."""
    return (
        item.get('name'),
        item.get('location'),
        item.get('zip'),
        item.get('type'),
    )

def forward_geocode(query: str):
    """Adresa → (lat, lon, ulice, mesto, psc, type). Vrátí tuple None při chybě."""
    try:
        r = requests.get(f'{MAPY_API}/geocode',
            params={'query': query, 'lang': 'cs', 'limit': 1, 'apikey': API_KEY},
            headers=HEADERS, timeout=10)
        r.raise_for_status()
        items = r.json().get('items', [])
        if not items:
            return None, None, None, None, None, None
        pos = items[0]['position']
        ulice, mesto, psc, typ = _parse_item(items[0])
        return pos['lat'], pos['lon'], ulice, mesto, psc, typ
    except Exception:
        return None, None, None, None, None, None

def reverse_geocode(lat, lon):
    """GPS → (ulice, mesto, psc, type).
    Preferuje regional.address, jinak první dostupný výsledek."""
    try:
        r = requests.get(f'{MAPY_API}/rgeocode',
            params={'lat': lat, 'lon': lon, 'lang': 'cs', 'apikey': API_KEY},
            headers=HEADERS, timeout=10)
        r.raise_for_status()
        items = r.json().get('items', [])
        if not items:
            return None, None, None, None
        item = next((i for i in items if i.get('type') == 'regional.address'), items[0])
        return _parse_item(item)
    except Exception:
        return None, None, None, None

print('Mapy.cz geocoding funkce načteny. Zkontroluj API_KEY.')

Mapy.cz geocoding funkce načteny. Zkontroluj API_KEY.


# ── Testovací buňka – rgeocode raw response ───────────────────────────────────
TEST_LAT = 50.0815534
TEST_LON = 14.4243770

r = requests.get(f'{MAPY_API}/rgeocode',
    params={'lat': TEST_LAT, 'lon': TEST_LON, 'lang': 'cs', 'apikey': API_KEY},
    headers=HEADERS, timeout=10)
r.raise_for_status()

items = r.json().get('items', [])
pd.DataFrame([
    {
        'type':     i.get('type'),
        'adresa':   i.get('name'),
        'mesto':    i.get('location'),
        'distance': i.get('distance'),
    }
    for i in items
])

def get_complete_data(lat, lon, api_key):
    url = "https://api.mapy.cz"
    params = {
        "lat": lat,
        "lon": lon,
        "apikey": api_key,
        "lang": "cs"
    }
    
    response = requests.get(url, params=params)
    
    print(f"HTTP Status: {response.status_code}")
    print(f"Obsah odpovědi (raw): {response.text}") # Toto ukáže, co se skutečně děje
    
    try:
        data = response.json()
        print("\n--- Strukturovaná data z API ---")
        for item in data.get("items", []):
            # Pro adresu s číslem popisným hledejte typ 'addr'
            print(f"Typ: {item.get('type'):<10} | Název: {item.get('name')}")
            # Vypíše všechny dostupné atributy pro každý záznam
            print(f"Detaily: {item}\n")
    except Exception as e:
        print(f"\nChyba při parsování JSONu: {e}")




# Nezapomeňte vložit svůj skutečný klíč z https://developer.mapy.cz/
MY_API_KEY = "nYJ2gE48fnO2TASz1wwFRCGYVrTfvi331EpqhdwWRvA"

# Souřadnice Hlavního nádraží (blízko Vrchlického sadů), kde je číslo popisné
latitude = 49.6888000
longitude = 14.0131000

get_complete_data(latitude, longitude, MY_API_KEY)



In [16]:
# Forward geocoding – adresa → GPS pro záznamy s flag_bez_gps = 1
# Vyplní gps_lat, gps_lon a zpětně doplní ulice_cp, mesto, psc z API odpovědi.

LIMIT_FORWARD = 100

to_forward = pd.read_sql(text('''
    SELECT id, typ_datasetu, ulice_cp, mesto, psc
    FROM   nemovitosti
    WHERE  gps_lat IS NULL
      AND  (ulice_cp IS NOT NULL OR mesto IS NOT NULL)
    LIMIT  :lim
'''), engine, params={'lim': LIMIT_FORWARD})

print(f'Forward geocoding: {len(to_forward)} záznamů')

with engine.begin() as conn:
    for _, row in to_forward.iterrows():
        parts = [to_none(row.get(c)) for c in ('ulice_cp', 'mesto', 'psc')]
        query = ', '.join(p for p in parts if p is not None)
        lat, lon, ulice, mesto, psc, typ = forward_geocode(query)
        conn.execute(text('''
            UPDATE nemovitosti
            SET gps_lat          = :lat,
                gps_lon          = :lon,
                ulice_cp         = COALESCE(:ulice, ulice_cp),
                mesto            = COALESCE(:mesto, mesto),
                psc              = COALESCE(:psc,   psc),
                geocoding_shoda  = :typ,
                geocoding_metoda = 'forward'
            WHERE id = :id AND typ_datasetu = :typ_ds
        '''), {'lat': lat, 'lon': lon, 'ulice': ulice, 'mesto': mesto,
               'psc': psc, 'typ': typ,
               'id': int(row['id']), 'typ_ds': row['typ_datasetu']})
        time.sleep(1)

print(f'Forward geocoding hotovo ({len(to_forward)} záznamů uloženo)')

Forward geocoding: 0 záznamů
Forward geocoding hotovo (0 záznamů uloženo)


In [17]:
# Reverse geocoding – GPS → adresa pro záznamy s prázdnou ulice_cp
# Vyplní ulice_cp (name z API), vždy přepíše mesto (location) a psc (zip).

LIMIT_REVERSE = 2000

to_reverse = pd.read_sql(text('''
    SELECT id, typ_datasetu, gps_lat, gps_lon
    FROM   nemovitosti
    WHERE  gps_lat IS NOT NULL
      AND  ulice_cp IS NULL
    LIMIT  :lim
'''), engine, params={'lim': LIMIT_REVERSE})

print(f'Reverse geocoding: {len(to_reverse)} záznamů')

with engine.begin() as conn:
    for _, row in to_reverse.iterrows():
        ulice, mesto, psc, typ = reverse_geocode(row['gps_lat'], row['gps_lon'])
        conn.execute(text('''
            UPDATE nemovitosti
            SET ulice_cp         = :ulice,
                mesto            = :mesto,
                psc              = :psc,
                geocoding_shoda  = :typ,
                geocoding_metoda = 'reverse'
            WHERE id = :id AND typ_datasetu = :typ_ds
        '''), {
            'ulice': ulice, 'mesto': mesto, 'psc': psc, 'typ': typ,
            'id': int(row['id']), 'typ_ds': row['typ_datasetu'],
        })
        time.sleep(1)

print(f'Reverse geocoding hotovo ({len(to_reverse)} záznamů uloženo)')

Reverse geocoding: 1185 záznamů
Reverse geocoding hotovo (1185 záznamů uloženo)
